### PR TITLE
Added K8s Route Filtering into wizard

### PR DESCRIPTION
**Pull request by hhovsepy**
RFE https://github.com/kiali/kiali/issues/5502

Added a new tab in K8s Gateway API Routing wizard: "Route Filtering".
It supports only "RequestHeaderModifier" type of filtering so far. The rest of filter types will be added in further PRs.

Screenshot from 2022-12-08 18-59-45
Screenshot from 2022-12-08 18-59-25

### CODE REVIEW

File: frontend/src/components/IstioWizards/K8sRequestRouting.tsx [hunk 1]
```diff
@@ -22,6 +23,11 @@ type State = {
   matchValue: string;
   k8sRules: K8sRule[];
   validationMsg: string;
+  headerOp: string;
+  headerValue: string;
+  filterType: string;
+  filterValue: string;
+  filters: string[];
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/K8sRequestRouting.tsx:30]: It's totally subjective but over the years I've found it quite useful to alphabetize these long lists of variable names, both when declaring, setting, etc.  Change is definitely not required (of course whenever modifying an existing list check to see if it's ordered, and maintain the ordering in that case).

File: frontend/src/components/IstioWizards/K8sRequestRouting.tsx [hunk 2]
```diff
@@ -163,10 +177,10 @@ class K8sRequestRouting extends React.Component<Props, State> {
 
   onHeaderNameChange = (headerName: string) => {
     let validationMsg = '';
-    if (this.state.matchValue !== '' && headerName === '') {
+    if ((this.state.matchValue !== '' || this.state.headerOp === REMOVE) && headerName === '') {
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/K8sRequestRouting.tsx:180]:
```suggestion
    if (!headerName && (!!this.state.matchValue || this.state.headerOp === REMOVE)) {
```

No change required but I think it's often better to use truthy/falsy to cover corner cases like `undefined` and such.

File: frontend/src/components/IstioWizards/K8sRequestRouting.tsx [hunk 3]
```diff
@@ -262,6 +276,56 @@ class K8sRequestRouting extends React.Component<Props, State> {
     }
   }
 
+  onAddFilter = () => {
+    this.setState(prevState => {
+      let newFilter: string;
+      if (this.state.filterType === REQ_MOD || this.state.filterType === RESP_MOD) {
+        if (this.state.headerOp !== REMOVE) {
+          newFilter = prevState.filterType + ' [' + prevState.headerName + '] ' + prevState.headerOp + ' ' + prevState.headerValue;
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/K8sRequestRouting.tsx:284]:
```suggestion
          newFilter = `${prevState.filterType} [${prevState.headerName}] ${prevState.headerOp} ${prevState.headerValue}`;
```

I can't find anything wrong in your code and so I'm just making subjective suggestions :-)  Again, no change required, but I find it less error prone to build strings in this way.

**Review by jshaughn — DISMISSED** [review]
Code looks good and seems to be working.  A couple of minor, non-blocking comments.   If there is a chance to add/update a cypress test then please do, but perhaps it is mainly covered already by non k8s api use case.

**Review by josunect — COMMENTED** [review]
Functionally it works fine, but if I try to create a filter with empty values, this is allowed in the GUI but is not allowed in the server validation: 

image

image

image

Server logs:

`2022-12-13T10:30:24Z ERR HTTPRoute.gateway.networking.k8s.io "details-v1" is invalid: [spec.rules[0].filters[0].requestHeaderModifier.add[0].value: Invalid value: "": spec.rules[0].filters[0].requestHeaderModifier.add[0].value in body should be at least 1 chars long, spec.rules[0].filters[0].requestHeaderModifier.set[1].value: Invalid value: "": spec.rules[0].filters[0].requestHeaderModifier.set[1].value in body should be at least 1 chars long, spec.rules[0].filters[0].requestHeaderModifier.set[1]: Duplicate value: map[string]interface {}{"name":"o"}]`

I guess it would be better to also perform this validation in the client side? 

It happens for "set" and "add".

File: frontend/src/components/IstioWizards/K8sRequestRouting.tsx [hunk 4]
```diff
@@ -176,6 +218,7 @@ class K8sRequestRouting extends React.Component<Props, State> {
   };
 
   onQueryParamNameChange = (queryParamName: string) => {
+    console.log("queryParamName " + queryParamName)
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/K8sRequestRouting.tsx:221]:
```suggestion
```

File: frontend/src/components/IstioWizards/K8sRequestRouting.tsx [hunk 5]
```diff
@@ -190,6 +233,7 @@ class K8sRequestRouting extends React.Component<Props, State> {
   };
 
   onMatchValueChange = (matchValue: string) => {
+    console.log("matchValue " + matchValue)
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/K8sRequestRouting.tsx:236]:
```suggestion
```

File: frontend/src/components/IstioWizards/K8sRequestRouting.tsx [hunk 6]
```diff
@@ -262,6 +304,79 @@ class K8sRequestRouting extends React.Component<Props, State> {
...
+      return {
+        filters: prevState.filters,
+        headerName: '',
+        headerValue: ''
+      };
+    });
+  };
+
+  onHeaderValueChange = (headerValue: string) => {
+    console.log("headerValue " + headerValue)
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/K8sRequestRouting.tsx:331]:
```suggestion
```

File: frontend/src/components/IstioWizards/K8sRequestRouting.tsx [hunk 7]
```diff
@@ -262,6 +304,79 @@ class K8sRequestRouting extends React.Component<Props, State> {
...
+      }
+    }
+    this.setState({
+      headerValue: headerValue,
+      validationMsg: validationMsg
+    });
+  }
+
+  onHostNameChange = (hostName: string) => {
+    console.log("hostName " + hostName)
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/K8sRequestRouting.tsx:348]:
```suggestion
```

File: frontend/src/components/IstioWizards/K8sRequestRouting.tsx [hunk 8]
```diff
@@ -262,6 +304,79 @@ class K8sRequestRouting extends React.Component<Props, State> {
...
+      validationMsg = MSG_HOSTNAME_NON_EMPTY;
+    }
+    this.setState({
+      hostName: hostName,
+      validationMsg: validationMsg
+    });
+  }
+
+  onPortValueChange = (portValue: string) => {
+    console.log("portValue " + portValue)
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/K8sRequestRouting.tsx:360]:
```suggestion
```

File: frontend/src/components/IstioWizards/K8sRequestRouting/K8sFilters.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  height: 40
+});
+
+const remove = style({
+  cursor: "not-allowed"
+});
+
+class K8sFilters extends React.Component<Props> {
+  render() {
+    console.log("this.props.filters " + JSON.stringify(this.props.filters))
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/K8sRequestRouting/K8sFilters.tsx:22]:
```suggestion
```

File: frontend/src/components/IstioWizards/WizardActions.ts [hunk 10]
```diff
@@ -353,6 +421,43 @@ const parseK8sHTTPMatchRequest = (httpRouteMatch: K8sHTTPRouteMatch): string[] =
   return matches;
 };
 
+const parseK8sHTTPRouteFilter = (httpRouteFilter: K8sHTTPRouteFilter): string[] => {
+  console.log("httpRouteFilter " + JSON.stringify(httpRouteFilter))
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/WizardActions.ts:425]:
```suggestion
```

File: frontend/src/components/IstioWizards/WizardActions.ts [hunk 11]
```diff
@@ -353,6 +421,43 @@ const parseK8sHTTPMatchRequest = (httpRouteMatch: K8sHTTPRouteMatch): string[] =
...
+  }
+  if (httpRouteFilter.requestRedirect) {
+    matches = matches.concat(parseK8sHTTPRouteRequestRedirect(httpRouteFilter.requestRedirect));
+  }
+  return matches;
+};
+
+const parseK8sHTTPHeaderFilter = (filterType: string, httpHeaderFilter: K8sHTTPHeaderFilter): string[] => {
+  const filters: string[] = [];
+  console.log("httpHeaderFilter " + JSON.stringify(httpHeaderFilter))
```
Comment (jshaughn) [inline frontend/src/components/IstioWizards/WizardActions.ts:438]:
```suggestion
```

**Review by josunect — COMMENTED** [review]
Testing the changes, I is possible to create a Route filter with no Header name/value, errors have been resolved. 

There are many logs in the console, but I can see they will be addressed in @jshaughn comments.

**Comment by hhovsepy** [comment]
oops, sorry for the noice with logs

**Comment by hhovsepy** [comment]
@jshaughn @josunect thanks for the review, console logs are removed now :)

**Review by jshaughn — APPROVED** [review]
Looks good to me, @hhovsepy !

**Review by josunect — APPROVED** [review]
Tested and works as expected, code LGTM. 

Thanks, @hhovsepy !

**Pull request merged**